In [1]:
from dataclasses import dataclass,field,asdict
import json
from typing import List, Dict
import models
import torch


In [2]:
from torch.utils.data import Dataset, DataLoader
import torchvision.transforms as T
import matplotlib.pyplot as plt
import numpy as np
from tqdm.notebook import tqdm
from utils.validation import get_validation_recalls
from models import helper
from utils import utils
import models

In [4]:
import warnings
warnings.filterwarnings("ignore")

# Тестирование

In [20]:
conf = utils.load_from_json('conf.json')
for test_name in conf.tests:
    for model_conf in conf.models:
        model_conf = utils.ModelItem(**model_conf)
        print(f'------- {model_conf.title} -------')
        model = models.MODELS_PULL[model_conf.model_name](**model_conf.model_conf)
        model.load_model_state_dict(model_conf.model_weight_path)
        device = torch.device('cuda:1')#model_conf.device)
        print(device)
        model.set_model_device(device)
        for dataset_name in conf.datasets:
            utils.run_test(model,dataset_name,test_name)

------- MixVPR-128 -------
cuda:1


Calculating descritptors...:   0%|          | 0/42 [00:00<?, ?it/s]

Descriptor dimension 128


+----------------------------------+
|       Performance on essex       |
+----------+-------+-------+-------+
|    K     |   1   |   5   |   10  |
+----------+-------+-------+-------+
| Recall@K | 82.86 | 93.81 | 98.10 |
+----------+-------+-------+-------+


Calculating descritptors...:   0%|          | 0/3036 [00:00<?, ?it/s]

Descriptor dimension 128


+----------------------------------+
|     Performance on nordland      |
+----------+-------+-------+-------+
|    K     |   1   |   5   |   10  |
+----------+-------+-------+-------+
| Recall@K | 32.43 | 50.07 | 56.78 |
+----------+-------+-------+-------+
------- MixVPR-512 -------
cuda:1


Calculating descritptors...:   0%|          | 0/42 [00:00<?, ?it/s]

Descriptor dimension 512


+-----------------------------------+
|        Performance on essex       |
+----------+-------+-------+--------+
|    K     |   1   |   5   |   10   |
+----------+-------+-------+--------+
| Recall@K | 87.62 | 99.52 | 100.00 |
+----------+-------+-------+--------+


Calculating descritptors...:   0%|          | 0/3036 [00:00<?, ?it/s]

Descriptor dimension 512


+----------------------------------+
|     Performance on nordland      |
+----------+-------+-------+-------+
|    K     |   1   |   5   |   10  |
+----------+-------+-------+-------+
| Recall@K | 44.17 | 60.58 | 67.03 |
+----------+-------+-------+-------+
------- MixVPR-4096 -------
cuda:1


Calculating descritptors...:   0%|          | 0/42 [00:00<?, ?it/s]

Descriptor dimension 4096


+------------------------------------+
|        Performance on essex        |
+----------+-------+--------+--------+
|    K     |   1   |   5    |   10   |
+----------+-------+--------+--------+
| Recall@K | 88.10 | 100.00 | 100.00 |
+----------+-------+--------+--------+


Calculating descritptors...:   0%|          | 0/3036 [00:00<?, ?it/s]

Descriptor dimension 4096


+----------------------------------+
|     Performance on nordland      |
+----------+-------+-------+-------+
|    K     |   1   |   5   |   10  |
+----------+-------+-------+-------+
| Recall@K | 55.11 | 70.33 | 75.47 |
+----------+-------+-------+-------+
------- BOQ VPR -------


Using cache found in /home/sunveil/.cache/torch/hub/facebookresearch_dinov2_main
/home/sunveil/.cache/torch/hub/facebookresearch_dinov2_main/dinov2/layers/swiglu_ffn.py:43: UserWarning: xFormers is available (SwiGLU)
  warnings.warn("xFormers is available (SwiGLU)")
/home/sunveil/.cache/torch/hub/facebookresearch_dinov2_main/dinov2/layers/attention.py:27: UserWarning: xFormers is available (Attention)
  warnings.warn("xFormers is available (Attention)")
/home/sunveil/.cache/torch/hub/facebookresearch_dinov2_main/dinov2/layers/block.py:33: UserWarning: xFormers is available (Block)
  warnings.warn("xFormers is available (Block)")


cuda:1


Calculating descritptors...:   0%|          | 0/42 [00:00<?, ?it/s]

Descriptor dimension 12288


+-----------------------------------+
|        Performance on essex       |
+----------+-------+-------+--------+
|    K     |   1   |   5   |   10   |
+----------+-------+-------+--------+
| Recall@K | 90.95 | 99.52 | 100.00 |
+----------+-------+-------+--------+


Calculating descritptors...:   0%|          | 0/3036 [00:00<?, ?it/s]

Descriptor dimension 12288


+----------------------------------+
|     Performance on nordland      |
+----------+-------+-------+-------+
|    K     |   1   |   5   |   10  |
+----------+-------+-------+-------+
| Recall@K | 81.49 | 92.39 | 94.75 |
+----------+-------+-------+-------+
------- VladBuff VPR -------


Using cache found in /home/sunveil/.cache/torch/hub/facebookresearch_dinov2_main
/media/sunveil/Data/header_detection/vpr-auto-tests/models/vladbuff_vpr.py:430: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues relate

Loaded model from ./saved_models/vlad_buff/dnv2_NV_AB_wpca8192_last.ckpt Successfully!
cuda:1


Calculating descritptors...:   0%|          | 0/42 [00:00<?, ?it/s]

Descriptor dimension 8192


+------------------------------------+
|        Performance on essex        |
+----------+-------+--------+--------+
|    K     |   1   |   5    |   10   |
+----------+-------+--------+--------+
| Recall@K | 91.90 | 100.00 | 100.00 |
+----------+-------+--------+--------+


Calculating descritptors...:   0%|          | 0/3036 [00:00<?, ?it/s]

Descriptor dimension 8192


+----------------------------------+
|     Performance on nordland      |
+----------+-------+-------+-------+
|    K     |   1   |   5   |   10  |
+----------+-------+-------+-------+
| Recall@K | 74.28 | 87.64 | 90.98 |
+----------+-------+-------+-------+


In [18]:
next(model.model.parameters()).is_cpu


True

In [19]:
model.load_model_state_dict(model_conf.model_weight_path)


/media/sunveil/Data/header_detection/vpr-auto-tests/models/vpr_interface.py:36: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  state_dict = torch.load(path_to_weight)


# Наборы данных

## ESSEX3IN1

![image.png](essex.jpeg)

Набор состоит из 420 фотографий (query = 210; reference = 210), представляет собой фотографии помещений, улиц и природных сцен.
Особенность набора в том, что он включает фотографии мест, которые являются запутанными как для VPR, так и для человеческого распознавания. Он содержит запутанные и сложные динамические объекты, естественные сцены и малоинформативные кадры. Как показано в нашей статье [«Запоминающиеся карты: A Framework for Re-defining Places in Visual Place Recognition»](https://arxiv.org/abs/1811.03529), большинство современных методов VPR с трудом справляются с этими запутанными изображениями.

Набор данных разделен на 2 папки. Соответствие между кадрами применяется к кадрам запроса и опорным кадрам. Изображения 0-132 в каждой папке являются запутанными, а изображения 133-209 - хорошими кадрами.
## Nordland

![image.png](nordland.jpg)

Набор состоит из 30352 фотографий (query = 2760; reference = 27592), представляет собой фотографии, собранные с поездов на железных дорогах Норвегии в разные времена суток и года. [Подробнее на ...](https://nrkbeta.no/2013/01/15/nordlandsbanen-minute-by-minute-season-by-season/)
## Irk
Иркутск(Академ городок) 2gis + yandex map
5412 изображений (285 query, 5127 ref) 329 Локаций (195 2gis, 134 yandex)

# Модели

### MixVPR
![Описание изображения](./mixvpr.png)

### Vlad-Buff VPR
![Описание изображения](./vladbuff.png)

### BOQ
![Описание изображения](./BOQ.jpg)

# Выводы


|    Model Name      |   Embedding dimension   | speed (img/sec) |
|:-----------------------------|:---------:|:---------:|
|            MIX              |  128/512/4096  | 560/560/540 |
|         VLADBUFF            |  8192  | 280 |
|            BOQ              |  12288  | 130 |




$$
Recall@k = \frac{Кол-во\ релевантных\ элементов\ в\ top-k}{Общее\ кол-во\ релевантных\ документов}
$$

Метрика используется, когда важно найти как можно больше релевантных элементов, даже если это означает включение некоторых нерелевантных.

### ESSEX
|    Recall@K on ESSEX      |   K=1   |   K=5   |   K=10  |
|-----------------------------|---------|---------|---------|
|            MIX 128             |   82.86 | 93.81 | 98.10   |
|            MIX 512             |  87.62 | 99.52 | *100.00*  |
|            MIX 4096             |  88.10  | *100.00*  | *100.00*  |
|         VLADBUFF            |  **91.90**  | *100.00*  | *100.00*  |
|            BOQ              |  90.95  |  99.52  | *100.00*  |
### NORDLAND
|    Recall@K on NORDLAND      |   K=1   |   K=5   |   K=10  |
|-----------------------------|---------|---------|---------|
|            MIX 128              |  32.43 | 50.07 | 56.78  |
|            MIX 512             |  44.17 | 60.58 | 67.03  |
|            MIX 4096             |  55.11 | 70.33 | 75.47  |
|         VLADBUFF            |  74.28 | 87.64 | 90.98  |
|            BOQ              |  **81.49** | **92.39** | **94.75**  |
### IRK
|    Recall@K on IRK      |   K=1   |   K=5   |   K=10  |
|-----------------------------|---------|---------|---------|
|            MIX 128             |   29.37 | 42.66 | 52.45  |
|            MIX 512             |  36.36 | 51.75 | 58.39  |
|            MIX 4096             |  38.46 | 56.99 | 61.54  |
|         VLADBUFF            |  44.76 | **67.13** | **72.73**  |
|            BOQ              |  **46.15** | 62.94 | 68.88  |
### Msls_Moscow
|    Recall@K on MSLS_Moscow  |   K=1      |   K=5     |   K=10     |
|-----------------------------|------------|-----------|------------|
|            MIX 128          |   58.73    | 67.44     | 71.14      |
|            MIX 512          |  63.78     | 71.27     | 74.25      |
|            MIX 4096         |  66.73     | 73.60     | 76.16      |
|               BOQ           |  **83.22** | **89.28** | **91.25**  |
|            VLADBUFF         |  78.30     |  85.36    | 87.69      |